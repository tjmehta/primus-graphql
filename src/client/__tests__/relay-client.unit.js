/* eslint-env jest */
const ExposedObservable = require('static-observable')
const Primus = require('primus')
const shimmer = require('shimmer')
const Subscriber = require('rxjs').Subscriber
require('rxjs/add/operator/publish')
require('observable-backoff')

const PrimusRelayClient = require('../relay-client')

// setup mocks
jest.unmock('../relay-client')
jest.mock('primus')

describe('PrimusRelayClient', () => {
  const ctx = {}

  beforeEach(() => {
    // stub console.error for silence
    shimmer.wrap(console, 'error', () => jest.fn())
    ctx.primus = new Primus() // mock
    ctx.primus.graphql = jest.fn()
    ctx.opts = {
      retry: {
        minTimeout: 0,
        retries: 1
      }
    }
  })
  afterEach(() => {
    // unstub console.error for silence
    shimmer.unwrap(console, 'error')
  })

  describe('constructor', () => {
    it('should create an instance', () => {
      expect(new PrimusRelayClient(ctx.primus, ctx.opts)).toBeInstanceOf(PrimusRelayClient)
    })
  })

  describe('methods', () => {
    beforeEach(() => {
      ctx.relayClient = new PrimusRelayClient(ctx.primus, ctx.opts)
    })

    describe('fetch', () => {
      describe('graphql success (200)', () => {
        beforeEach(() => {
          ctx.payload = {
            statusCode: 200
          }
          ctx.primus.graphql.mockImplementation(() => Promise.resolve(ctx.payload))
        })

        it('should send graphql query', () => {
          const operation = { text: 'operationText' }
          const variables = { var1: 'foo', var2: 'bar' }
          return ctx.relayClient.fetch(operation, variables).then((payload) => {
            expect(ctx.primus.graphql).toMatchSnapshot()
            expect(payload).toBe(ctx.payload)
          })
        })
      })

      describe('graphql client error (unknown)', () => {
        beforeEach(() => {
          ctx.primus.graphql.mockImplementation(() => Promise.reject(new Error('network error')))
        })

        it('should send graphql query', () => {
          const operation = { text: 'operationText' }
          const variables = { var1: 'foo', var2: 'bar' }
          return ctx.relayClient.fetch(operation, variables).then(() => {
            throw new Error('this should not happen')
          }).catch((err) => {
            expect(err).toMatchSnapshot()
          })
        })
      })

      describe('graphql non-retryable error (400)', () => {
        beforeEach(() => {
          ctx.payload = {
            statusCode: 400
          }
          ctx.primus.graphql.mockImplementation(() => Promise.resolve(ctx.payload))
        })

        it('should send graphql query', () => {
          const operation = { text: 'operationText' }
          const variables = { var1: 'foo', var2: 'bar' }
          return ctx.relayClient.fetch(operation, variables).then(() => {
            throw new Error('this should not happen')
          }).catch((err) => {
            expect(err).toMatchSnapshot()
          })
        })
      })

      describe('graphql retryable error (500)', () => {
        beforeEach(() => {
          ctx.payload = {
            statusCode: 500
          }
          ctx.primus.graphql.mockImplementation(() => Promise.resolve(ctx.payload))
        })

        it('should send graphql query', () => {
          const operation = { text: 'operationText' }
          const variables = { var1: 'foo', var2: 'bar' }
          return ctx.relayClient.fetch(operation, variables).then(() => {
            throw new Error('this should not happen')
          }).catch((err) => {
            expect(err).toMatchSnapshot()
          })
        })
      })

      describe('timeout error', () => {
        beforeEach(() => {
          ctx.opts = {
            timeout: 0,
            retry: {
              minTimeout: 0,
              retries: 0
            }
          }
          ctx.relayClient = new PrimusRelayClient(ctx.primus, ctx.opts)
          ctx.primus.graphql.mockImplementation(() => new Promise(() => {}))
        })

        it('should yield timeout error', () => {
          const operation = { text: 'operationText' }
          const variables = { var1: 'foo', var2: 'bar' }
          // fetch
          const promise = ctx.relayClient.fetch(operation, variables)
          // assertions
          return promise.then(() => {
            throw new Error('this should not happen')
          }).catch((err) => {
            expect(err).toMatchSnapshot()
          })
        })

        describe('1 retry', () => {
          beforeEach(() => {
            ctx.opts.retry.retries = 1
            ctx.relayClient = new PrimusRelayClient(ctx.primus, ctx.opts)
            ctx.primus.graphql.mockImplementation(() => new Promise(() => {}))
          })
          it('should yield timeout error', () => {
            const operation = { text: 'operationText' }
            const variables = { var1: 'foo', var2: 'bar' }
            // fetch
            const promise = ctx.relayClient.fetch(operation, variables)
            // assertions
            return promise.then(() => {
              throw new Error('this should not happen')
            }).catch((err) => {
              expect(err).toMatchSnapshot()
            })
          })
        })
      })
    })

    describe('subscribe', () => {
      beforeEach(() => {
        ctx.observable = new ExposedObservable()
        ctx.primus.graphql.mockImplementation(() => ctx.observable)
        ctx.observer = {
          onCompleted: jest.fn(),
          onError: jest.fn(),
          onNext: jest.fn()
        }
      })

      it('should subscribe to graphql subscription', () => {
        const operation = { text: 'operationText' }
        const variables = { var1: 'foo', var2: 'bar' }
        const cacheConfig = null
        const subscriber = ctx.relayClient.subscribe(operation, variables, cacheConfig, ctx.observer)
        expect(subscriber).toBeInstanceOf(Subscriber)
        expect(subscriber.unsubscribe).toEqual(expect.any(Function))
      })

      it('should subscribe to subscription and retry on error (short error)', (done) => {
        const operation = { text: 'operationText' }
        const variables = { var1: 'foo', var2: 'bar' }
        const cacheConfig = null
        ctx.relayClient.subscribe(operation, variables, cacheConfig, ctx.observer)
        // short error w/ no stack for coverage
        const err = new Error('')
        delete err.stack
        ctx.observable.error(err)
        setTimeout(() => {
          expect(ctx.observer).toMatchSnapshot()
          done()
        }, 100)
      })

      it('should subscribe to subscription and retry on error (long error)', (done) => {
        const operation = { text: 'operationText' }
        const variables = { var1: 'foo', var2: 'bar' }
        const cacheConfig = null
        ctx.relayClient.subscribe(operation, variables, cacheConfig, ctx.observer)
        // long error w/ no stack for coverage
        const err = new Error('long message')
        err.status = 500
        delete err.stack
        ctx.observable.error(err)
        setTimeout(() => {
          expect(ctx.observer).toMatchSnapshot()
          done()
        }, 100)
      })

      describe('subscribed', () => {
        beforeEach(() => {
          ctx.operation = { text: 'operationText' }
          ctx.variables = { var1: 'foo', var2: 'bar' }
          ctx.cacheConfig = null
          ctx.subscriber = ctx.relayClient.subscribe(ctx.operation, ctx.variables, ctx.cacheConfig, ctx.observer)
        })

        it('should recieve next', (done) => {
          const nextData = 'nextData'
          ctx.observable.next(nextData)
          setTimeout(() => {
            expect(ctx.observer).toMatchSnapshot()
            done()
          }, 100)
        })
      })
    })
  })
})
