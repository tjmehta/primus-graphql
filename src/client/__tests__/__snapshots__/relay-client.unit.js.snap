// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimusRelayClient methods fetch graphql client error (unknown) should send graphql query 1`] = `[Error: fetch(): GraphQL client error, tried 2 times. ]`;

exports[`PrimusRelayClient methods fetch graphql non-retryable error (400) should send graphql query 1`] = `[Error: fetch(): Received error response from server, tried 1 times. ]`;

exports[`PrimusRelayClient methods fetch graphql retryable error (500) should send graphql query 1`] = `[Error: fetch(): Received error response from server, tried 2 times. ]`;

exports[`PrimusRelayClient methods fetch graphql success (200) should send graphql query 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "operationText",
      Object {
        "var1": "foo",
        "var2": "bar",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`PrimusRelayClient methods fetch timeout error 1 retry should yield timeout error 1`] = `[Error: fetch(): Request timed out, tried 2 times. ]`;

exports[`PrimusRelayClient methods fetch timeout error should yield timeout error 1`] = `[Error: fetch(): Request timed out, tried 1 times. ]`;

exports[`PrimusRelayClient methods subscribe should subscribe to subscription and retry on error (long error) 1`] = `
Object {
  "onCompleted": [MockFunction],
  "onError": [MockFunction] {
    "calls": Array [
      Array [
        [Error: long message],
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": undefined,
      },
    ],
  },
  "onNext": [MockFunction],
}
`;

exports[`PrimusRelayClient methods subscribe should subscribe to subscription and retry on error (short error) 1`] = `
Object {
  "onCompleted": [MockFunction],
  "onError": [MockFunction] {
    "calls": Array [
      Array [
        [Error],
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": undefined,
      },
    ],
  },
  "onNext": [MockFunction],
}
`;

exports[`PrimusRelayClient methods subscribe subscribed should recieve next 1`] = `
Object {
  "onCompleted": [MockFunction],
  "onError": [MockFunction],
  "onNext": [MockFunction] {
    "calls": Array [
      Array [
        Object {
          "data": "nextData",
        },
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": undefined,
      },
    ],
  },
}
`;
